Fix autocomplete

LTangaF opened the pull request:
Closes #36 

Let me know if this needs to be a PR against the dev branch instead.
Also, the protocol was a guess.

LTangaF committed:
Add compiler override, re-using nil>>compiler

LTangaF committed:
Enable auto-complete from remote side

LTangaF commented:
Maybe hold off on merge. I was doing this development in Pharo 8. I should test it in Pharo 7.

dionisiydk commented on TelePharo-Playground.package/TlpRemoteDoItReceiver.class/instance/compiler.st:
```diff
@@ -0,0 +1,4 @@
+accessing
+compiler
+	"Shameless hack to get at the CompilerContext of the remote system for help with remote CompletionContext"
+	^ self remotePharo evaluate: [ nil compiler ]
```
Could you explain why we need a compiler from the remote side? 
Can we use "self class compiler" instead?

One of ideas was to be able to work with very tiny images which does not have even a compiler.
But I appreciate that in current state that could be a simplest fix for you problem

MarcusDenker replied:
What is important is to get a compiler for the class that you look at. Else it can not highlight instance variables or class variables.

LTangaF replied:
Thank you both for the feedback. Initially, this method wasn't present and was the source of the stack trace in Pharo 8 and I did stumble across `self class compilerClass` while trying to solve the more advanced issue of the auto-complete context. If you use a local compiler on the client side, it doesn't get the `environment` (a Dictionary of all the classes and methods) from the remote side.

My test for this was to create a class only on the remote side and try to get it to show up during auto-complete.

In my exploration of standard, non-TelePharo behavior of auto-complete in the Playground, it wasn't actually binding to a specific class compiler, but using the compiler of the nil class, so that's why I thought it was probably safe to do so.  I'm thinking that perhaps if you're working in the context of an editor _IN_ the browser, it may possibly bind to the class you're actually working on, but that wasn't one of the things I was testing.

dionisiydk replied:
> If you use a local compiler on the client side, it doesn't get the `environment` (a Dictionary of all the classes and methods) from the remote side.

Yes. The trick to set the remote environment (available in #remotePharo instance) into the local compiler. It is how the remote inspector implements remote doIts.
We can merge your version and experiment later but maybe you can try it at this step

dionisiydk commented:
Currently RemoteDoItReceiver is specifically for playground where nil in
context breaks the "remote hook". In the inspector it would be a proxy and
"everything would just works".
So I introduced this special object to address tis problem.

But generally the integration with tools definitely requires a more love.

чт, 28 мая 2020 г. в 09:49, Marcus Denker <notifications@github.com>:

> *@MarcusDenker* commented on this pull request.
> ------------------------------
>
> In
> TelePharo-Playground.package/TlpRemoteDoItReceiver.class/instance/compiler.st
> <https://github.com/pharo-ide/TelePharo/pull/37#discussion_r431680266>:
>
> > @@ -0,0 +1,4 @@
> +accessing
> +compiler
> +	"Shameless hack to get at the CompilerContext of the remote system for help with remote CompletionContext"
> +	^ self remotePharo evaluate: [ nil compiler ]
>
> What is important is to get a compiler for the class that you look at.
> Else it can not highlight instance variables or class variables.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pharo-ide/TelePharo/pull/37#discussion_r431680266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6FVIBJ3L5WH47XSXFGTOLRTYQSDANCNFSM4NMXMAZA>
> .
>

LTangaF commented:
@dionisiydk I can't tell if that's an endorsement of what I did or if your trying to guide me to a better solution (which guidance I would be happy to have).

LTangaF commented:
Confirmed that Pharo 7 isn't touching this new code. The TlpRemoteDoItReceiver is still registered as the metaclass for the editor, but the popup dispatch code is different in Pharo 7 from Pharo 8.

The Inspect It or Do It hot keys interestingly work fine, while the autocomplete context doesn't in Pharo 7. It looks like the latter will be a lot more of a chore to fix.

dionisiydk commented:
> The Inspect It or Do It hot keys interestingly work fine, while the autocomplete context doesn't in Pharo 7. It looks like the latter will be a lot more of a chore to fix.

We probably have some worse effect from code completion in remote case. Generally editors (in the browser especially) are quite slow. They produce too many remote requests.

> @dionisiydk I can't tell if that's an endorsement of what I did or if your trying to guide me to a better solution (which guidance I would be happy to have).

The main problem with TelePharo now is the editor. @oliveiraallex is trying to fix this. But I do not know a progress. 
The idea is to disable all obvious parts (completion, highlighting, shortcuts), prove that they are responsible for the bad behavior and enable them step by step to find the main issue. 

The main tool here is a Seamless profiler which you can try using:
```Smalltalk
SeamlessLogger startAFresh.
"do anything in the image"
SeamlessLogger collectStatistics inspect
```
It will show statistics over all remote messages grouped by classes and selectors

LTangaF commented:
My first commit just handled the missing method in a very rudimentary and local way which solved the stack trace issue without trying to solve the code completion issue.

I read some comments in the code about that slow behavior, so I was fully expecting that my attempt and resolving the code completion issue was going to result in that and that I was going to maybe have to find some cool way of marshaling up the full class and method list (basically the contents of the environment Dictionary) in a single request to populate the local compiler proxy or something in order to deal with that potential slow-down.

The performance issue that was anticipated was not realized, though, and I think it may be because that the narrowWith: may be being handled completely on the remote side with this solution.

Anyway, take a look at the first commit and I could revert to just that to resolve the stack trace issue and let @oliveiraallex work on the longer term solution.

dionisiydk commented:
> I read some comments in the code about that slow behavior, so I was fully expecting that my attempt and resolving the code completion issue was going to result in that and that I was going to maybe have to find some cool way of marshaling up the full class and method list (basically the contents of the environment Dictionary) in a single request to populate the local compiler proxy or something in order to deal with that potential slow-down.

it's interesting problem how to make remote completion fast. Would it require special design decisions to facilitate this property? How transparent it could be? (to use same local completion in remote context). And consider that ideally remote side should have minimum support for that.

Transferring full environment dictionary does not scale. It is a lot of classes. Sending 8000 objects is too slow to be executed interactively. Completion queries should be evaluated on remote side and only the required (visible) part of result should be sent to the client. This is how a browser works: when you query remote senders only visible part of senders are loaded and it is done in single request.

LTangaF commented:
Maybe I can explore a little more to find out what's taking place. My sense is that the compilationContext and narrowWith operations may actually be occurring on the remote side, but I have to figure out how to investigate.  When I do `CTRL-I` on the result of that `TlpRemotePharo evaluate: [ nil compiler ]` in the RPlayground, since that is still broken I actually get an Inspector on a SeamlessProxy object and not a remote inspector on an OpalCompiler. I'm thinking that being fully proxied, the code execution for that may not be transferring 8000 objects. It seemed fairly snappy.

That said, I could do further investigation before you merge if you wanted. I just need to figure out how. BTW, does that environment Dictionary contain 8000 full objects or just 8000 symbol objects?

LTangaF commented:
I feel pretty confident now that the building of the list that populates the context popup menu occurs only on the remote side.

   - I identified the list creation occurring in RBVariableNode>>completionEntries: (provided as an extention in the NECompletion package).
   - I tested out a local breakpoint at the top of that package with the TelePharo client disconnected in a normal Playground.
   - I then connected TelePharo and RPlayground and started typing in that. The breakpoint wasn't hit.
   - I removed the breakpoint and tried to add a breakpoint in the remote Browser, but breakpoint setting is broken in the proxies right now.
   - I disconnected TelePharo and went onto the remote Pharo image and added the breakpoint to the same method on that side and reconnected TelePharo client
   - I tried typing in the new RPlayground and it froze. Nothing was visible on the remote image. I waited just a few and used `ALT-.` and both the interrupt debugger and the breakpoint debugger appeared. The breakpoint debugger in the remote session was sitting at the right place for the breakpoint
   - The stack was full of SeamlessProxy items, but since it was the remote Browser, I think maybe they were the SeamlessProxy objects pointing back from the remote to the local

I think I can conclude with some confidence that it's not pulling back the full environment dictionary to do this list creation locally.

LTangaF commented:
Oh. And I could probably add the #hack flag to the method if desired.

LTangaF commented:
Thanks. I didn't bother trying to solve the autocomplete thing in Pharo 7 (as I mentioned above) because it looks like it's wired in in a way that would make that more challenging. I did confirm that the commits here in this PR had no effect on Pharo 7, though.

dionisiydk commented:
> BTW, does that environment Dictionary contain 8000 full objects or just 8000 symbol objects?

Forget to answer. 
This environment is Smalltalk globals. It's all classes in the system (and other globals). I guess you know that. 
Counting the internal structure of the dictionary will give at least three times more objects: internal dictionary associations +  key (class name) and class itself (and internal array to hold all of them). 
Also classes will be transferred by reference (proxies on the client side) and that would require maintaining same amount of objects on the server side (server has a map reference->object). Of course it can be specialized and optimized for classes but the default current behavior is like that.

LTangaF commented:
That makes sense (and is now obvious now that it's pointed out to me). I had asked that before I confirmed that it wasn't sending that across the link.
